pkg/actuators/machine/instances.go

```diff
@@ -363,68 +376,142 @@ func getImageID(machine runtimeclient.ObjectKey, machineProviderConfig *alibabac
 	klog.Infof("%s validate security group in region %s", machineProviderConfig.SecurityGroupID, machineProviderConfig.RegionID)
+	var securityGroupIDs []string
+
+	// If SecurityGroupID is assigned, use it directly
+	if machineProviderConfig.SecurityGroupID != "" {
+		securityGroupIDs = append(securityGroupIDs, machineProviderConfig.SecurityGroupID)
+	} else {
+		// Otherwise, the query securityGroupIDs by the tags
+		for _, sg := range machineProviderConfig.SecurityGroups {
+			if sg.ID != "" {
+				securityGroupIDs = append(securityGroupIDs, sg.ID)
+			} else {
+				if sg.Tags != nil {
+					request := ecs.CreateDescribeSecurityGroupsRequest()
+					if machineProviderConfig.VpcID != "" {
+						request.VpcId = machineProviderConfig.VpcID
+					}
+					if machineProviderConfig.ResourceGroupID != "" {
+						request.ResourceGroupId = machineProviderConfig.ResourceGroupID
+					}
+					request.RegionId = machineProviderConfig.RegionID
+					request.Tag = buildDescribeSecurityGroupsTag(sg.Tags)
+					request.Scheme = "https"
+
+					response, err := client.DescribeSecurityGroups(request)
+					if err != nil {
+						metrics.RegisterFailedInstanceCreate(&metrics.MachineLabels{
+							Name:      machine.Name,
+							Namespace: machine.Namespace,
+							Reason:    err.Error(),
+						})
+						klog.Errorf("error describing securitygroup: %v", err)
+						return nil, fmt.Errorf("error describing securitygroup: %v", err)
+					}
+
+					if len(response.SecurityGroups.SecurityGroup) < 1 {
+						klog.Errorf("no securitygroup for given tags not found")
+						return nil, fmt.Errorf("no securitygroup for given tags not found")
+					}
+
+					for _, sg := range response.SecurityGroups.SecurityGroup {
+						securityGroupIDs = append(securityGroupIDs, sg.SecurityGroupId)
+					}
+				}
+			}
+		}
+	}
 
-	request := ecs.CreateDescribeSecurityGroupsRequest()
-	request.VpcId = machineProviderConfig.VpcID
-	request.RegionId = machineProviderConfig.RegionID
-	request.SecurityGroupId = machineProviderConfig.SecurityGroupID
-	request.Scheme = "https"
+	return &securityGroupIDs, nil
+}
 
-	response, err := client.DescribeSecurityGroups(request)
-	if err != nil {
-		metrics.RegisterFailedInstanceCreate(&metrics.MachineLabels{
-			Name:      machine.Name,
-			Namespace: machine.Namespace,
-			Reason:    err.Error(),
-		})
-		klog.Errorf("error describing securitygroup: %v", err)
-		return "", fmt.Errorf("error describing securitygroup: %v", err)
+func getMaxInstancesBySecurityGroupType(securityGroupType string) int {
+	switch securityGroupType {
+	case SecurityGroupTypeNoraml:
+		return MaxInstanceOfSecurityGroupTypeNoraml
+	case SecurityGroupTypeEnterprise:
+		return MaxInstanceOfSecurityGroupTypeEnterprise
+	default:
+		return MaxInstanceOfSecurityGroupTypeNoraml
 	}
+}
+
+func buildDescribeSecurityGroupsTag(tags []alibabacloudproviderv1.Tag) *[]ecs.DescribeSecurityGroupsTag {
```

Comment:
This function returns a group of 6 security group tags. 3 empty ones and 3 correct ones which causes the query to fail.

Suggest fixing by using an index and storing them as you did with vswitch tags here:
https://github.com/openshift/cluster-api-provider-alibaba/pull/9/files#diff-c5079a7c5d2916c3661d0c44d1267050b3496c40b4530363d3825d66af581b6fR507-R508